ProjectUtils.waitForProjects() not enough to wait for markers to appear/disappear, leading to flaky tests

We have a few places where we test our validation puts problem markers properly, e.g.,

- Tests in `XmlValidatorTest` such as `testXsdValidation_appengineWebXml()` and `testValidate_badXml()`
- [`XsltSourceQuickFixTest.testApply()`](https://github.com/GoogleCloudPlatform/google-cloud-eclipse/blob/v1.2.0B/plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/XsltSourceQuickFixTest.java#L86)
- `ToServlet25SourceQuickFixTest.testConvertServlet()`

We call `ProjectUtils.waitForProjects()` to give some time for the markers to be updated. However, it seems like we need to wait further, as the tests occasionally fail: https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/2075#issuecomment-309565629 and https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/2069#issuecomment-309488235.

This issue does not necessarily mean we want to overload `waitForProjects()` further to also cover marker updates. We could put some code to wait for markers in the tests themselves.

```
Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 25.888 sec <<< FAILURE! - in com.google.cloud.tools.eclipse.appengine.validation.XsltSourceQuickFixTest
testApply(com.google.cloud.tools.eclipse.appengine.validation.XsltSourceQuickFixTest)  Time elapsed: 25.687 sec  <<< FAILURE!
java.lang.AssertionError: [/test0.06296464251836487/appengine-web.xml:2: Project ID should be specified at deploy time]
	at org.junit.Assert.fail(Assert.java:88)
	at com.google.cloud.tools.eclipse.test.util.ArrayAssertions.assertIsEmpty(ArrayAssertions.java:58)
	at com.google.cloud.tools.eclipse.test.util.ArrayAssertions.assertIsEmpty(ArrayAssertions.java:36)
	at com.google.cloud.tools.eclipse.appengine.validation.XsltSourceQuickFixTest.testApply(XsltSourceQuickFixTest.java:92)
```
It's like the old markers are being removed independently?

Happened a bit earlier in `XsltSourceQuickFixTest.testApply()`, even before making any change to the XML. This suggests this flaky marker failure have nothing to do with the XML editor.

```
testApply(com.google.cloud.tools.eclipse.appengine.validation.XsltSourceQuickFixTest)  Time elapsed: 3.767 sec  <<< FAILURE!
java.lang.AssertionError: expected:<1> but was:<0>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.google.cloud.tools.eclipse.appengine.validation.XsltSourceQuickFixTest.testApply(XsltSourceQuickFixTest.java:76)
```

--- a/plugins/com.google.cloud.tools.eclipse.appengine.validation.test/META-INF/MANIFEST.MF
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.validation.test/META-INF/MANIFEST.MF
@@ -14,6 +14,8 @@ Import-Package: com.google.cloud.tools.eclipse.test.util,
  org.eclipse.core.runtime.jobs,
  org.eclipse.jst.common.project.facet.core,
  org.eclipse.jst.j2ee.web.project.facet,
+ org.eclipse.swt.widgets,
  org.junit;version="4.12.0",
+ org.hamcrest;version="1.1.0",
  org.mockito;provider=google;version="1.10.19",
  org.mockito.stubbing;provider=google;version="1.10.19"

--- a/plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/XsltSourceQuickFixTest.java
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/XsltSourceQuickFixTest.java
@@ -68,9 +68,11 @@ public class XsltSourceQuickFixTest {
     IWorkbench workbench = PlatformUI.getWorkbench();
     IEditorPart editorPart = WorkbenchUtil.openInEditor(workbench, file);
     ITextViewer viewer = ValidationTestUtils.getViewer(file);
-    String preContents = viewer.getDocument().get();
-    workbench.getDisplay().readAndDispatch(); // spin the event loop
+    while(workbench.getDisplay().readAndDispatch()) {
+      // spin the event loop
+    }
 
+    String preContents = viewer.getDocument().get();
     assertThat(preContents, containsString("application"));
 
     ProjectUtils.waitForProjects(project);

--- a/plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/XsltSourceQuickFixTest.java
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/XsltSourceQuickFixTest.java
@@ -16,10 +16,10 @@
 
 package com.google.cloud.tools.eclipse.appengine.validation;
 
-import static org.junit.Assert.assertArrayEquals;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.not;
 import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.assertThat;
 
 import com.google.cloud.tools.eclipse.appengine.facets.AppEngineStandardFacet;
 import com.google.cloud.tools.eclipse.test.util.ArrayAssertions;
@@ -69,8 +69,9 @@ public class XsltSourceQuickFixTest {
     IEditorPart editorPart = WorkbenchUtil.openInEditor(workbench, file);
     ITextViewer viewer = ValidationTestUtils.getViewer(file);
     String preContents = viewer.getDocument().get();
+    workbench.getDisplay().readAndDispatch(); // spin the event loop
 
-    assertTrue(preContents.contains("application"));
+    assertThat(preContents, containsString("application"));
 
     ProjectUtils.waitForProjects(project);
     assertEquals(1, file.findMarkers(BLACKLIST_MARKER, true, IResource.DEPTH_ZERO).length);
@@ -81,8 +82,8 @@ public class XsltSourceQuickFixTest {
 
     IDocument document = viewer.getDocument();
     String contents = document.get();
-    assertFalse(contents.contains("application"));
-    assertFalse(contents.contains("?><appengine"));
+    assertThat(contents, not(containsString("application")));
+    assertThat(contents, not(containsString("?><appengine")));
 
     // https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/1527
     editorPart.doSave(new NullProgressMonitor());

